### PR TITLE
Softmax refactor

### PR DESCRIPTION
**Pull request by wentingj**
## Description ##
1. refact softmax op with signature
2. encapsulate mkldnn operation into class

### Essentials ###
- [ ] Passed code style checking (`make lint`)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] For user-facing API changes, API doc string has been updated. For new C++ functions in header files, their functionalities and arguments are well-documented. 
- [ ] To my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: src/operator/nn/mkldnn/mkldnn_softmax.cc [hunk 1]
```diff
@@ -21,34 +21,109 @@
...
+
+ private:
+  void _Init(const SoftmaxParam& param,
+             bool is_train,
+             const NDArray &in_data,
+             const NDArray &out_data,
+             const OpReqType &req) {
+    // mkldnn::softmax_forward::primitive_desc
+    auto input_mem = in_data.GetMKLDNNData();
+    auto output_mem = out_data.GetMKLDNNData();
```
Comment (TaoLv) [inline src/operator/nn/mkldnn/mkldnn_softmax.cc]: Calling GetMKLDNNData() will create memory for out_data with default layout. But out_data may need a specific layout which is same as in_data?

File: src/operator/nn/mkldnn/mkldnn_softmax.cc [hunk 2]
```diff
@@ -21,34 +21,109 @@
...
+
+    mkldnn::memory::primitive_desc data_mpd = input_mem->get_primitive_desc();
+    mkldnn::memory::desc data_md = data_mpd.desc();
+    auto cpu_engine = data_mpd.get_engine();
```
Comment (TaoLv) [inline src/operator/nn/mkldnn/mkldnn_softmax.cc]: CpuEngine::Get()->get_engine()

File: src/operator/nn/mkldnn/mkldnn_softmax.cc [hunk 3]
```diff
@@ -21,34 +21,109 @@
...
+    stream->RegisterPrim(*fwd);
+    stream->Submit();
+  }
+
+ private:
+  void _Init(const SoftmaxParam& param,
+             bool is_train,
+             const NDArray &in_data,
+             const NDArray &out_data,
+             const OpReqType &req) {
```
Comment (TaoLv) [inline src/operator/nn/mkldnn/mkldnn_softmax.cc:67]: req is not used in this function?

File: src/operator/nn/mkldnn/mkldnn_softmax.cc [hunk 4]
```diff
@@ -21,34 +21,109 @@
...
+                   const NDArray &in_data,
+                   const NDArray &out_data,
+                   const OpReqType &req):
+                   fwd(nullptr), data(nullptr), out(nullptr) {
+                   _Init(param, is_train, in_data, out_data, req);
+  }
+  ~MKLDNNSoftmaxFwd() {}
+  void SetDataHandle(const NDArray &in_data,
+                     const NDArray &out_data,
+                     const OpReqType &req) {
```
Comment (TaoLv) [inline src/operator/nn/mkldnn/mkldnn_softmax.cc]: req is not used in this function.

Comment (wentingj) [reply src/operator/nn/mkldnn/mkldnn_softmax.cc]: yes, I will remove this input.

File: src/operator/nn/mkldnn/mkldnn_softmax.cc (hunk 1 above)
Comment (zheng-da) [reply src/operator/nn/mkldnn/mkldnn_softmax.cc]: correct. This is a problem in my original code.

File: src/operator/nn/softmax.cc [hunk 5]
```diff
@@ -62,8 +62,9 @@ inline static bool SoftmaxStorageType(const nnvm::NodeAttrs& attrs,
   CHECK_EQ(out_attrs->size(), 1);
 
 #if MXNET_USE_MKLDNN == 1
-  // We only run MKLDNN op if it runs on CPU.
-  if (dev_mask == mshadow::cpu::kDevMask)
+  // We only run MKLDNN op if it runs on CPU and the input data is MKLDNN
+  // format.
+  if (dev_mask == mshadow::cpu::kDevMask && (*in_attrs)[0] == kMKLDNNStorage)
```
Comment (zheng-da) [inline src/operator/nn/softmax.cc]: Remove kMKLDNNStorage. There isn't kMKLDNNStorage any more.

File: src/operator/nn/mkldnn/mkldnn_softmax.cc [hunk 6]
```diff
@@ -21,34 +21,109 @@
...
+static MKLDNNSoftmaxFwd &GetSoftmaxFwd(const SoftmaxParam& param,
+                                       const OpContext &ctx,
+                                       const NDArray &in_data,
+                                       const NDArray &out_data,
+                                       const OpReqType &req) {
+  static thread_local std::unordered_map<MKLDNNSmSignature, MKLDNNSoftmaxFwd, MKLDNNOpHash> fwds;
+  MKLDNNSmSignature key(param);
+  key.AddSign(ctx.is_train);
+  key.AddSign(param.axis);
+  key.AddSign(in_data);
```
Comment (zheng-da) [inline src/operator/nn/mkldnn/mkldnn_softmax.cc]: We should add MKLDMM memory as part of a signature. We need layout info.

File: src/operator/nn/mkldnn/mkldnn_softmax.cc [hunk 7]
```diff
@@ -68,11 +67,10 @@ class MKLDNNSoftmaxFwd {
              const OpReqType &req) {
     // mkldnn::softmax_forward::primitive_desc
     auto input_mem = in_data.GetMKLDNNData();
-    auto output_mem = out_data.GetMKLDNNData();
-
     mkldnn::memory::primitive_desc data_mpd = input_mem->get_primitive_desc();
+    auto output_mem = CreateMKLDNNMem(out_data, data_mpd, req).second;
```
Comment (zheng-da) [inline src/operator/nn/mkldnn/mkldnn_softmax.cc]: if you use CreateMKLDNNMem, you have to use it with CommitOutput. You can't use CreateMKLDNNMem alone.
You could use the CreateMKLDNNData method in NDArray, as batchnorm does.

**Comment by wentingj** [comment]
softmax has its unittest already, and mkldnn can pass: nosetests tests/python/unittest/test_operator.py:test_new_softmax
